adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -0,0 +1,978 @@
+        final URL authorityUrl = StringExtensions.getUrl(authenticationRequest.getAuthority());
+        if (authorityUrl == null) {
+            callbackHandle.onError(new AuthenticationException(
+                    ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL));
+            return null;
+        }
+
+        if (validateAuthority && !mAuthorityValidated) {
+            // Discovery call creates an Async Task to send
+            // Web Requests
+            // using a handler
+            boolean result = validateAuthority(authorityUrl, authenticationRequest.getCorrelationId());
+            if (result) {
+                mAuthorityValidated = true;
+                Logger.v(TAG, "Authority is validated: " + authorityUrl.toString());
+            } else {
+                Logger.v(TAG, "Call external callback since instance is invalid"
+                        + authorityUrl.toString());
+                callbackHandle.onError(new AuthenticationException(
+                        ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_INSTANCE));
+                return null;
+            }
+        }
+
+        // Validated the authority or skipped the validation
+        try {
+            return acquireTokenAfterValidation(callbackHandle, activity, useDialog, authenticationRequest);
+        } catch (final AuthenticationException authenticationException) {
+            callbackHandle.onError(authenticationException);
+            return null;
+        }
+    }
+
+    /**
+     * Perform authority validation.
+     * True if the passed in authority is valid, false otherwise.
+     */
+    private boolean validateAuthority(final URL authorityUrl, final UUID correlationId) {
+        // This is not calling outer callback. It is using
+        // authenticationCallback, so handler is not needed here
+        if (mDiscovery != null) {
+            Logger.v(TAG, "Start validating authority");
+
+            // Set CorrelationId for Instance Discovery
+            mDiscovery.setCorrelationId(correlationId);
+            boolean result = mDiscovery.isValidAuthority(authorityUrl);
+            Logger.v(TAG, "Finish validating authority:" + authorityUrl + " result:" + result);
+            return result;
+        }
+
+        return false;
+    }
+
+    /**
+     * 1. For Silent flow, we should always try to look local cache first.
+     *    i> If valid AT is returned from cache, use it.
+     *    ii> If no valid AT is returned, but RT is returned, use the RT.
+     *    iii> If RT request fails, and if we can talk to broker, go to broker and check if there is a valid token.
+     * 2. For Non-Silent flow.
+     *    i> Do silent cache lookup first, same as 1.
+     *       a) If we can talk to broker, go to broker for auth.
+     *       b) If not, launch webview with embedded flow.
+     */
+    private AuthenticationResult acquireTokenAfterValidation(CallbackHandler callbackHandle,
```

Comment:
Document the three states that could return. 

---
In reply to: [66171490](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/655#discussion_r66171490) [](ancestors = 66171490)